src/ray/core_worker/experimental_channel_manager.cc

```diff
@@ -0,0 +1,200 @@
+#endif
+  return Status::OK();
+}
+
+bool ExperimentalChannelManager::ReaderChannelRegistered(
+    const ObjectID &object_id) const {
+  return reader_channels_.count(object_id);
+}
+
+Status ExperimentalChannelManager::ReadAcquire(const ObjectID &object_id,
+                                               std::shared_ptr<RayObject> *result) {
+#ifdef __linux__
+  auto reader_channel_entry = reader_channels_.find(object_id);
+  if (reader_channel_entry == reader_channels_.end()) {
+    return Status::ObjectNotFound("Reader channel has not been registered");
+  }
+  auto &channel = reader_channel_entry->second;
+  RAY_RETURN_NOT_OK(EnsureGetAcquired(channel));
+
+  RAY_CHECK(static_cast<int64_t>(channel.mutable_object->header->data_size +
+                                 channel.mutable_object->header->metadata_size) <=
+            channel.mutable_object->allocated_size);
+  auto data_buf = SharedMemoryBuffer::Slice(
+      channel.mutable_object->buffer, 0, channel.mutable_object->header->data_size);
+  auto metadata_buf =
+      SharedMemoryBuffer::Slice(channel.mutable_object->buffer,
+                                channel.mutable_object->header->data_size,
+                                channel.mutable_object->header->metadata_size);
+
+  *result = std::make_shared<RayObject>(
+      std::move(data_buf), std::move(metadata_buf), std::vector<rpc::ObjectReference>());
+
+#endif
+  return Status::OK();
+}
+
+Status ExperimentalChannelManager::ReadRelease(const ObjectID &object_id) {
+#ifdef __linux__
+  auto reader_channel_entry = reader_channels_.find(object_id);
+  if (reader_channel_entry == reader_channels_.end()) {
+    return Status::ObjectNotFound("Reader channel has not been registered");
+  }
+
+  auto &channel = reader_channel_entry->second;
+  RAY_RETURN_NOT_OK(EnsureGetAcquired(channel));
+
+  RAY_RETURN_NOT_OK(
+      channel.mutable_object->header->ReadRelease(channel.next_version_to_read));
+  // The next read needs to read at least this version.
+  channel.next_version_to_read++;
+  channel.read_acquired = false;
+
+#endif
+  return Status::OK();
+}
+
+Status ExperimentalChannelManager::EnsureGetAcquired(ReaderChannel &channel) {
+#ifdef __linux__
+  if (channel.read_acquired) {
+    return Status::OK();
+  }
+
+  int64_t version_read = 0;
+  RAY_RETURN_NOT_OK(channel.mutable_object->header->ReadAcquire(
+      channel.next_version_to_read, &version_read));
+  RAY_CHECK(version_read > 0);
+  channel.next_version_to_read = version_read;
+  channel.read_acquired = true;
+#endif
+  return Status::OK();
+}
+
+Status ExperimentalChannelManager::SetError(const ObjectID &object_id) {
+#ifdef __linux__
+  PlasmaObjectHeader *header = nullptr;
+  auto reader_channel_entry = reader_channels_.find(object_id);
+  if (reader_channel_entry != reader_channels_.end()) {
+    header = reader_channel_entry->second.mutable_object->header;
+  } else {
+    auto writer_channel_entry = writer_channels_.find(object_id);
```

Comment:
is there any case we have both reader/writer registered for the same object at the same time?

Reply:
Yes, this can happen although it shouldn't if you're using DAGs.

Reply:
Should we disallow it in this case?